Tasks/Common/dotNetCliWrapper.ts

```diff
@@ -112,27 +111,94 @@ export class DotNetCliWrapper {
         args.push('--disable-interactive')
         args.push('true')
 
-        return this.executeAsync(args, additionalArgs)
+        return await this.execute(args, additionalArgs)
     }
 
-    public async executeAsync(args: string[], additionalArgs: string): Promise<void> {
-        const dotnetPath = tl.which('dotnet', true)
-        console.log('Path to tool: ' + dotnetPath)
-
-        const dotnet = tl.tool(dotnetPath)
+    public async execute(args: string[], additionalArgs: string, additionalCommandLineOptions?: any): Promise<number> {
+        const dotnet = tl.tool(this.dotnetCliPath)
 
         for (const arg of args) {
             dotnet.arg(arg)
         }
 
         dotnet.line(additionalArgs)
 
-        const execOptions = {
+        const execOptions: any = {
             cwd: this.cwd,
             env: this.env
         }
 
+        if (additionalCommandLineOptions) {
+            // tslint:disable-next-line: no-unsafe-any
+            for (const key of Object.keys(additionalCommandLineOptions)) {
+                // tslint:disable-next-line: no-unsafe-any
+                execOptions.key = additionalCommandLineOptions[key]
+            }
+        }
+
         // tslint:disable-next-line: no-unsafe-any
-        await dotnet.exec(execOptions as any)
+        return await dotnet.exec(execOptions)
+    }
+
+    private async restore(): Promise<number> {
+        return await this.execute(['restore'], '')
+    }
+
+    private async checkForGlobalLambdaToolsInstalled(): Promise<boolean> {
+        try {
+            const returnCode = await this.execute(['lambda', 'help'], '', { silent: true })
+            if (returnCode !== 0) {
+                return false
+            }
+        } catch (exception) {
+            return false
+        }
+
+        return true
+    }
+
+    private async installGlobalTools(): Promise<boolean> {
+        try {
+            const returnCode = await this.execute(['tool', 'install', '-g', 'Amazon.Lambda.Tools'], '')
+            if (returnCode === 0) {
+                return true
+            }
+        } catch (e) {}
+
+        // If something went wrong in the last step, we try to update the tools instead
+        try {
+            const returnCode = await this.execute(['tool', 'update', '-g', 'Amazon.Lambda.Tools'], '')
+            if (returnCode === 0) {
+                return true
+            }
+        } catch (e2) {
+            return false
+        }
+
+        return false
+    }
+
+    public static async buildDotNetCliWrapper(cwd: string, env: any, dotnetCliPath: string): Promise<DotNetCliWrapper> {
+        const wrapper = new DotNetCliWrapper(cwd, env, dotnetCliPath)
+        tl.debug(tl.loc('StartingDotNetRestore'))
+        // Restore, this will install the lambda cli if it's < version 3
+        await wrapper.restore()
+        if (await wrapper.checkForGlobalLambdaToolsInstalled()) {
```

Comment:
Does this check prevent the tool from updating the tool if it was already installed by a previous run of the task and we have shipped a new version?

Reply:
Yes, are the tools guaranteed to be backwards compatible and not breaking? If so, we can update them here, otherwise there is potential for breakage. Hosted machines (most of our customers) will always have the latest version installed, so this should only affect build agents owned by customers.

Reply:
Yes we have the requirement of backwards compatible. You already have that requirement for the lambda tooling because you will grab the latest one if it is a fresh install. The code just isn't always updating so it puts us in an inconsistent state.

Also if there ever was a problem with Amazon.Lambda.Tools by not allowing updates we would not be able to ship a fix to Amazon.Lambda.Tools that would be picked up by this task.

Reply:
Sounds good added update